libr/io/submap.c

```diff
@@ -0,0 +1,289 @@
+	sm->to = _map_to (map);
+	sm->id = map->id;
+	r_io_submap_cut_out (io, sm->from, sm->to);
+	io->submaps->cmp = _submap_insert_no_overlap_cb;
+	ut32 od;
+	r_oids_insert (io->submaps, sm, &sm->sid, &od, NULL);
+}
+
+//cuts out all submaps, so that there is no submap between from and to
+R_API bool r_io_submap_cut_out(RIO *io, ut64 from, ut64 to) {
+	RIOSubMap *bd, *sm;
+	ut32 od;
+	if (to < from) {
+		if (!r_io_submap_cut_out (io, from, UT64_MAX)) {
+			return false;
+		}
+		from = 0LL;
+	}
+	if (!io || !io->submaps) {
+		return false;
+	}
+	if (!io->submaps->ptop) {
+		return true;
+	}
+	io->submaps->cmp = _submap_find_cb;
+	od = r_oids_find (io->submaps, &from, NULL);
+	while ((sm = r_oids_oget (io->submaps, od))) {
+//sm->to is guaranteed to be bigger or equal to from,
+//because that's how r_oids_find works
+		if (sm->from < from) {
+			if (sm->to > to) {
+//      ##cut this out##
+//############sm##############
+//becomes
+//      ##cut this out##
+//##sm##                ##bd##
+				bd = R_NEW (RIOSubMap);
+				if (!bd) {
+					return false;
+				}
+				bd->from = to + 1;
+				bd->to = sm->to;
+				bd->id = sm->id;
+				sm->to = from - 1;
+				io->submaps->cmp = _submap_insert_no_overlap_cb;
+				r_oids_insert (io->submaps, bd, &bd->sid, &od, NULL);
+				return true;
+			}
+//    ##cut this out##
+//###sm###
+//becomes
+//    ##cut this out##
+//#sm#
+			sm->to = from - 1;
+			od++;
+			continue;
+		}
+		if (sm->to <= to) {
+//     ##cut this out##
+//          ####sm####
+//sm gets deleted
+			free (sm);
+			r_oids_odelete (io->submaps, od);
+			continue;
+		}
+		if (sm->from <= to) {
+//  ##cut this out##
+//            ####sm####
+//becomes
+//  ##cut this out##
+//                  #sm#
+			sm->from = to + 1;
+//any further submap is going to be bigger than sm,
+//no need to continue the loop
+		}
+		return true;
+	}
+	return true;
+}
+
+//in - incoming
+static bool _submap_hack_n_slash_insert(void *vin, void *vincoming, void *user, int *cmp_res) {
+	RIOSubMap *in = (RIOSubMap *)vin;
+	RIOSubMap *incoming = (RIOSubMap *)vincoming;
+	RQueue *todo = (RQueue *)user;
+	RIOSubMap *map;
+#if 0
+	      ########in#########
+#incoming#
+#endif
+	if (in->from > incoming->to) {
+		cmp_res[0] = 1;
+		return true;
+	}
+#if 0
+########in#########
+			#incoming#
+#endif
+	if (in->to < incoming->from) {
+		cmp_res[0] = -1;
+		return true;
+	}
+	if (in->from > incoming->from) {
+#if 0
+	########in#########
+##incoming##
+#endif
+		if (in->to >= incoming->to) {
+			//cut off the intersecting part
+			incoming->to = in->from - 1;
+			cmp_res[0] = 1;
+			return true;
+		} else {
+#if 0
+	########in#########
+#############incoming###############
+#endif
+			//slash submap into 2 submaps, that don't intersect with the existing submap in
+			map = R_NEW0 (RIOSubMap);	//check needed
+			map->id = incoming->id;
+			map->to = incoming->to;
+			map->from = in->to + 1;
+			r_queue_enqueue (todo, map);	//will be inserted on the later iterations
```

Comment:
The queue used here makes reasoning about the logic difficult. The existing skyline is simpler.